Custom Margin/Padding to column block

AlexChariyska opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/master/docs/contributors/repository-management.md#pull-requests. -->

## Description
#24874 Custom functionality for adding margin and padding spacing in column/s blocks.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

## Screenshots <!-- if applicable -->
![image](https://user-images.githubusercontent.com/6911590/95577137-4364ea00-0a3a-11eb-86ce-2a66a120c969.png)


## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->
New Feature, created a hook for custom margin added in SpacingPanelControl.
Adds support for spacing in column/s block json.

## How to test:
In order to enable the custom padding functionality in **_lib/experimental-default-theme.json_** you have to set to **true** the customPadding key: 
`"spacing": { "customPadding": true }`

Currently, the custom margin and padding functionality are added only to the columns/column blocks.

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

achariyska committed:
Custom Margin/Padding to column block

youknowriad commented:
This is a very nice PR as it's the first one to potentially introduce the margin support flag that can be expanded to several blocks later.

I'd love a lot of eyes on this one: @jasmussen @ItsJonQ @nosolosw

jasmussen commented:
Thank you for the PR, exciting!

I'm afraid I can't get the custom spacing to work, honeslty anywhere, not on the Cover block either, and I'm opting in using `add_theme_support('custom-spacing');` — any idea what I'm missing, Riad?

This is what I see in the mean time, and I suppose it can serve as a "before" screenshot:

<img width="1474" alt="Screenshot 2020-10-12 at 13 07 59" src="https://user-images.githubusercontent.com/1204802/95740244-71da0380-0c8c-11eb-8454-1fc64d2d914f.png">

ItsJonQ commented:
@jasmussen I believe you have to example it via an `experimental-theme.json` file :). At least, that's the setup I have.

@AlexChariyska Thank you for attempting this!
I gave a spin, and the controls appear 👍 . There are somethings worth noting:

### ✨ Visualizers

The blue visualizers (that appear for `padding`) aren't working for margin. This is due to how they (the visualizers) are styled. I think we'll need to make some adjustments so that we can see them for `margin` adjustments.

### 📏 `margin: auto`

I attempted to add some margin on a Cover block. This broke the alignment of things. Initially, I wasn't sure why... After poking around Chrome's inspector tools, I realized it was because the horizontal margin was set to `0`, which normally would be considered neutral/default. However, it looks like Cover (and perhaps others) are relying on `auto` margin for left/right centering.

The BoxControl UI currently doesn't accommodate values other than number * unit values (e.g. `10%`).

We'll need to look into updating the design + logic to support this.

achariyska committed:
Adding visualizers for margin

AlexChariyska commented:
So I updated the request and added visualizers for the margin props as well.

Example margin:
![Screenshot 2020-10-14 at 16 19 26](https://user-images.githubusercontent.com/6911590/95996125-fff3ec80-0e3a-11eb-8a15-95c2bb4e219a.png)

And padding:
![Screenshot 2020-10-14 at 16 15 53](https://user-images.githubusercontent.com/6911590/95996133-02eedd00-0e3b-11eb-9dc6-bcd7564ca426.png)

The changes include refactoring of the BoxControlVisualizer to accept object of spacing values (margin & padding) in order to decide which borders to visualize for the block.

About the second part to accept  more custom values (ex. margin: auto) I think something like the custom relative units can be done as a design & logic ([25825](https://github.com/WordPress/gutenberg/pull/25825)).

achariyska committed:
Visualizer for inner blocks

kirilzh committed:
remove padding visualizer flickering

kirilzh committed:
fix spread values

kirilzh committed:
extract styles to a separate configuration object

AlexChariyska committed:
Allow auto margin unit

AlexChariyska committed:
Fix spreading of values

ItsJonQ commented:
Adding support for the `auto` in the dropdown will be tricky.
I've been experimenting with [something similar](https://github.com/ItsJonQ/g2/issues/66).

## 🤔 Trickiness

At the moment, the `UnitControl` parses units by splitting a `string` into `number` + `unit` (unit), like:

`10px` => `[10, 'px']`

This does not allow for string values like `auto`.

It also does not do [CSS style validation](https://developer.mozilla.org/en-US/docs/Web/API/CSSStyleValue/parse). I've realized that this is **incredibly** important - especially to have it built into the control.

For example, a valid value for `margin` may be `auto`, or `0` (unitless).

(Not taking into consideration values like `calc()`)

Thankfully, I was able to [come up with ideas](https://github.com/ItsJonQ/g2/issues/66) on how we may tackle this!

## 🪓 Patch Current UnitControl

We could adjust the `UnitControl` unit to allow for `auto` for this feature.
An update that taps into the unit parsing "step", adding a condition for `auto` somehow.
(I can help with this if needed 🙏 )

Long term, we need a more robust solution 👍

AlexChariyska committed:
Patch Current UnitControl

AlexChariyska committed:
Allow only margin to have an auto value

AlexChariyska commented:
@ItsJonQ Great idea 👍, it will be fantastic to have more control in the unit and better support for more custom values.
In the meantime, as suggested, with @kirilzh we implemented a patch for the UnitControl to accept an auto value.

kirilzh committed:
unit validation

ItsJonQ commented:
@AlexChariyska Haii! Wow! Thanks for making the update to add `auto` support.

I took it for a spin, and it looks like adding values via the (input) controls are working, for both Padding and Margin (including `auto` values).

I'm not able to see the visualizers though (for either one).
I'll dig a little deeper this afternoon to see if I can figure it out 🤔 

Just leaving this update now, in case it helps with debugging/development :)

ItsJonQ commented on packages/components/src/box-control/all-input-control.js:
```diff
@@ -16,9 +16,11 @@ export default function AllInputControl( {
+	const [ singleValue ] = allValues.slice( 0, 1 );
 	const hasValues = isValuesDefined( values );
 	const isMixed = hasValues && isValuesMixed( values );
+	const disableUnits = isMixed || singleValue?.match( /^(auto|calc)/ );
```
Very cool work-around!

retrofox commented on .gitignore:
```diff
@@ -22,3 +22,5 @@ test/native/junit.xml
 # Local overrides
 .wp-env.override.json
+
+.idea
```
a minor one: is this something that we want to add here?

kirilzh replied:
it's because IntelliJ's WebStorm keeps config settings in there. We removed it

paaljoachim commented:
Here is a similar PR for the Group block.

It would be nice to move padding and margin out of the experimental status making it available for a larger testing base. Making it available in the Gutenberg plugin.
(At the moment I do not know how to test this PR. Even with the above guidance I still do not get it.)

paaljoachim commented:
I believe that through this PR that we will have a general UI method of adding padding and margin controls to various blocks. I believe Dave @getdave closed the #16730 PR as the general method is being worked on here in this PR. It can then be transferred over to the Group block.

kirilzh committed:
fixed dragging on spacing units

kirilzh commented:
@ItsJonQ we updated the input fields and it should work as expected now.

ItsJonQ commented:
Hey all!

I just tested out the latest. I'm still not able to get the Cover margin/padding settings working on the initial entry.
I've recorded a quick screencast highlighting the interactions:

https://d.pr/v/eoiUB2

In it, I...

* Add a brand new Cover block
* Click on `Padding`
* Press <kbd>Up</kbd> on the keyboard to increment up
* Cover (in canvas) doesn't update.
* I do the same thing with `Margin`
* Then try dragging to update
* Cover (in canvas) doesn't update.
* I try individual values.
* Cover (in canvas) still doesn't update.
* I swap the unit value for the `Padding`.
* Cover (in canvas) does it's first update.

Hope this helps with debugging 🙏

paaljoachim commented:
How should the default padding between the column blocks be handled?

![Screen Shot 2020-11-15 at 18 35 28](https://user-images.githubusercontent.com/5323259/99192131-72752680-2771-11eb-937c-f7371d5ffd00.png)

For instance if I wanted to remove the padding between red and blue I would today have to go in and through CSS remove the padding. 

.wp-block-column:not(:first-child) {
    margin-left: 2em; <----- I would have to remove this to also remove the padding between the columns.
}

kirilzh commented:
@ItsJonQ Hey, I'm not sure what might be causing the problem. I'm following the steps you described and I can't reproduce the problem. Can you check again if you are up to date with `AlexChariyska:add/24874-add-margin-and-padding-to-columns`
![Screen-Recording-2020-11-18-at-1(1)](https://user-images.githubusercontent.com/17088041/99514171-1f2eee00-2994-11eb-8199-2da25dbba8bc.gif)

@paaljoachim I think it's ok to leave it as is because it provides a point of reference. You can return to the default state by resetting the values. The blocks will also move to the edges of the view if you unlink the values and change each individually.

kirilzh committed:
fixed unlink on default state

ItsJonQ commented:
@kirilzh Haiii! Apologies for the delayed response! Happy New Year :). I hope you're doing well.

I just pulled the latest from this PR. I'm still having issues with the controls.
I went through the flow that was demonstrated in the GIF

Here's a screen recording:
https://d.pr/v/nXXm2D

AlexChariyska committed:
fix: Initial value not set up

ItsJonQ commented:
@AlexChariyska Haii! Your latest update seemed to do the trick for me. The margin/padding values are being applied as expected now 🎉

stokesman commented:
Great work here @AlexChariyska and @kirilzh. I made #28451 based off of this. One thing I came across is that trying to zero the values doesn't seem to work. 

https://user-images.githubusercontent.com/9000376/105653729-61334f00-5e71-11eb-8a60-2ab99bdd4c66.mp4

I am in favor of this feature. I think I'd like it even more if it supported negative values and that would enable it to support use cases such as #24504. Though it might be a trick to figure out what to do with the Visualizer for such cases.

I do think this has some potential as a pitfall for some users. It probably requires an understanding of margin collapsing and auto margins in order for some of the behavior to be expected.

paaljoachim commented:
If I remember correctly the Spacing panel is being renamed to Dimensions. @ItsJonQ 
Should we also add a Gutter setting into the dimensions panel as well?

Mamaduka commented:
The spacing support (currently only padding) was added to the Column block in #31778.
